go/vt/vtgate/evalengine/fn_string.go

```diff
@@ -433,3 +447,191 @@ func (call *builtinWeightString) compile(c *compiler) (ctype, error) {
 		return ctype{Type: sqltypes.VarBinary, Flag: flagNullable | flagNull, Col: collationBinary}, nil
 	}
 }
+
+func (call builtinLeftRight) eval(env *ExpressionEnv) (eval, error) {
+	str, l, err := call.arg2(env)
+	if err != nil {
+		return nil, err
+	}
+	if str == nil || l == nil {
+		return nil, nil
+	}
+
+	text, ok := str.(*evalBytes)
+	if !ok {
+		text, err = evalToVarchar(str, call.collate, true)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	length := evalToInt64(l).i
+	if length <= 0 {
+		return newEvalText(nil, text.col), nil
+	}
+
+	cs := text.col.Collation.Get().Charset()
+	strLen := charset.Length(cs, text.bytes)
+
+	if strLen <= int(length) {
+		return newEvalText(text.bytes, text.col), nil
+	}
+
+	var res []byte
+	if call.left {
+		res = charset.Slice(cs, text.bytes, 0, int(length))
+	} else {
+		res = charset.Slice(cs, text.bytes, strLen-int(length), strLen)
+	}
+	return newEvalText(res, text.col), nil
+}
+
+func (call builtinLeftRight) typeof(env *ExpressionEnv, fields []*querypb.Field) (sqltypes.Type, typeFlag) {
+	_, f1 := call.Arguments[0].typeof(env, fields)
+	return sqltypes.VarChar, f1
+}
+
+func (call builtinLeftRight) compile(c *compiler) (ctype, error) {
+	str, err := call.Arguments[0].compile(c)
+	if err != nil {
+		return ctype{}, err
+	}
+
+	l, err := call.Arguments[1].compile(c)
+	if err != nil {
+		return ctype{}, err
+	}
+
+	skip := c.compileNullCheck2(str, l)
+
+	col := defaultCoercionCollation(c.cfg.Collation)
+	switch {
+	case str.isTextual():
+		col = str.Col
+	default:
+		c.asm.Convert_xc(2, sqltypes.VarChar, col.Collation, 0, false)
+	}
+	_ = c.compileToInt64(l, 1)
```

Comment:
It's not clear to me what this does here since it doesn't seem to mutate `l`:
```
func (c *compiler) compileToInt64(ct ctype, offset int) ctype {
	switch ct.Type {
	case sqltypes.Int64:
		return ct
	case sqltypes.Uint64:
		c.asm.Convert_ui(offset)
	// TODO: specialization
	default:
		c.asm.Convert_xi(offset)
	}
	return ctype{sqltypes.Int64, ct.Flag, collationNumeric}
}
```

 We could generally use a few more comments in this code. :-) Comment could clear it up.

Reply:
@mattlord That is correct since `l` is only the type for the current expression. This is in general how the compiler works. It emits the code to actually do the conversion at runtime, it doesn't modify the variable `l`, but it compiles a method that ends up converting the value at the offset `1`. 

I don't think a single comment here would help clarify that really, because this is more a fundamental way how the compiler works. It might more indicate a usefulness for some generic description about how the compiler works instead?